proxy_agent/src/data_vessel.rs

```
@@ -0,0 +1,99 @@
+            let mut cached_key: Key = Key::empty(); // start with empyt key
+
+            while let Ok(action) = receiver.recv() {
+                match action {
+                    DataAction::Stop => {
+                        break;
+                    }
+                    DataAction::KeyKeeperSetKey { key, response } => {
+                        // Set the key from the key keeper
+                        cached_key = key.clone();
+                        _ = response.send(true);
+                    }
+                    DataAction::KeyKeeperGetKeyValue { response } => {
+                        // Get the key from the key keeper
+                        _ = response.send(cached_key.clone());
+                    }
+                }
+            }
+        });
+
+        DataVessel { sender }
+    }
+
+    pub fn stop(&self) {
+        let _ = self.sender.send(DataAction::Stop);
+    }
+}
+
+/// KeyKeeper implementation
+impl DataVessel {
+    pub fn update_current_key(&self, key: Key) -> bool {
+        let (response, receiver) = std::sync::mpsc::channel::<bool>();
+        let _ = self
+            .sender
+            .send(DataAction::KeyKeeperSetKey { key, response });
+        match receiver.recv() {
+            Ok(result) => result,
+            Err(e) => {
+                logger::write_warning(format!("Failed to update current key with error: {e}"));
+                false
+            }
+        }
+    }
+
+    fn get_current_key(&self) -> Key {
+        let (response, receiver) = std::sync::mpsc::channel::<Key>();
+        let _ = self
+            .sender
+            .clone()
+            .send(DataAction::KeyKeeperGetKeyValue { response });
+        match receiver.recv() {
+            Ok(key) => key,
+            Err(e) => {
+                logger::write_warning(format!("Failed to get current key with error: {e}"));
+                // return empty key if failed to get the key
+                Key::empty()
```

You misunderstood me, it's not that the `unwrap_or` approach wouldn't work, it's that making up a key if the actor can't be reached instead of returning an error seems incorrect.

---

if it could not be reached, it logs and returns an invalid empty key. The caller should able to handle from here.

---

This should return an error instead of hiding it behind an valid-looking invalid value.